.pre-commit-hooks.yaml

```diff
@@ -0,0 +1,15 @@
+- id: pylint-ignore
+  name: pylint
+  entry: pylint-ignore
+  language: python
+  types: [python]
+  require_serial: true
+
+- id: pylint-ignore-update
+  name: update pylint-ignore file
+  entry: pylint-ignore
+  args: ["--update-ignorefile"]
+  stages: [manual]
+  language: python
+  types: [python]
+  require_serial: true
```

Comment:
I am very wary of this hook. I think you'll agree that it should only be used as a one off thing, just as the command `pylint-ignore --update-ignorefile` should only be used as a one off thing.

My understanding is, that the purpose of this hook is to bring the execution environments in line with each other (and therefore generate the same pylint-ginore.md). If this is so, could we recommend instead, to the use the configuration option `ignore-paths=scripts/,fixtures/,setup.py`. This way `pylint` checks the same set of files, no matter the environment, so we wouldn't need the extra hook and we

Reply:
We've been having some interesting experiences with `pylint` and `pylint-ignore` over the last couple of weeks, related to this, which has led to the delay in my replying. We have three environments in which `pylint-ignore` runs:

1. run on the command line in the development environment on specific files or directories,
1. run by `pre-commit` in the development environment on changed files in a commit or all Python files in the repository (via `git ls`) and
1. run by `pre-commit` in a Docker container on all Python files in the repository.

`pylint` complicates this situation because we've observed that the rule violations reported depend on both the environment _and_ the set of Python files passed as arguments. Any rules which perform checks across the file set, e.g. R0801 duplicate-code, will be sensitive to this. It may also be the case that the file arguments are provided in a different order and that `pylint`'s output is non-deterministic.

What this hook is attempting to achieve is the creation of a `pylint-ignore.md` file which is the union of the sets of rule violations in (1) and (2). The problem we encountered is that _additional_ rule violations were reported in (3) which causes the lint job in the pipeline to fail. This appears to be related to differences between the development environment (macOS; Windows 10 with [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/)) and the CI environment (Ubuntu 20.04 running in Docker). In one instance that I was able to resolve, it was a version difference in a sub-dependency which sorted lists.

Unfortunately, on one project using MicroPython we were continually running into "whack-a-mole" situations where changes to the code were causing (2) and (3) to report violations which weren't reported in (1). The last straw was some issue causing (3) to report violations which I wasn't able to resolve in a reasonable time, and on this project we reverted to not using `pylint` and `pylint-ignore`. The intention is to re-enable this week, after refactoring the repository and _by necessity_ regenerating the pylint-ignore.md file, and try again.

The problem of `pylint` reporting different rule violations depending on the context in which it's run would seem to have a deeper cause than just the set of files passed as arguments. It seems that some rules are incompatible with the requirement for consistent output between (1-3), but it also seems that working out which rules are  incompatible could be like playing "whack-a-mole" again.

I'm not sure how to proceed from this point, it would seem to require a much greater understanding of Python, `pylint` and their relationship to MicroPython than we presently have.